### PR TITLE
storage: enforce sorted traversal during Walk

### PR DESCRIPTION
**Pull request by stevvooe**
This ensures sorted order of elements encountered during walk. This open up a number of optimizations for the catalog API, in addition to other enumeration methods.

The error propagation for Walk has also been fixed.

Attempts to address issues in #1190.

### CODE REVIEW

File: registry/storage/walk.go [hunk 1]
```diff
@@ -26,7 +27,12 @@ func Walk(ctx context.Context, driver storageDriver.StorageDriver, from string,
 	if err != nil {
 		return err
 	}
+	sort.Stable(sort.StringSlice(children))
```
Comment (pdevine) [inline registry/storage/walk.go]: Maybe a comment here about the expected cost of calling sort?

Comment (stevvooe) [inline registry/storage/walk.go]: Cost is here: https://godoc.org/sort#Stable. Click through to code to see the full cost analysis. I don't think we'll hit asymptotic bounds very often.

**Comment by pdevine** [comment]
LGTM.

**Comment by RichardScothern** [comment]
LGTM

**Pull request merged**